test/associative/ferguson/check-hashing-statistics.chpl

```diff
@@ -0,0 +1,254 @@
+  msg[10] = (e & 0xffffffff  ):uint(32);
+  msg[11] = (f >> 32         ):uint(32);
+  msg[12] = (f & 0xffffffff  ):uint(32);
+  msg[13] = (g >> 32         ):uint(32);
+  msg[14] = (g & 0xffffffff  ):uint(32);
+  msg[15] = (h >> 32         ):uint(32);
+  msg[16] = (h & 0xffffffff  ):uint(32);
+
+  var hash:8*uint(32) = state.lastblock(msg, 16*32);
+  var ret: uint(64);
+  ret = hash(1);
+  ret <<= 32;
+  ret |= hash(2);
+  return ret;
+}
+
+
+proc shrink(x:uint) {
+  const Muint = M:uint;
+  if useTop then
+    return (x >> 58):int;
+  else
+    return (x%Muint):int;
+}
+
+proc doHash(x:int) {
+  if useSHA {
+    return sha256hash(x:uint);
+  } else {
+    return chpl__defaultHash(x);
+  }
+}
+
+proc doCombine(a, b, n) {
+  if useSHA {
+    return sha256hash(a:uint, b:uint, n:uint);
+  } else {
+    return chpl__defaultHashCombine(a, b, n);
+  }
+}
+
+proc printHisto(counts) {
+  for (i, count) in zip(counts.domain, counts) {
+    if count != 0 then
+      writef("[% 8i] = %i\n", i, count);
+  }
+}
+
+proc summarize(histo) {
+  var theMin = max(int);
+  var theMax = min(int);
+  var theAvg = 0;
+  for count in histo {
+    if count < theMin then
+      theMin = count;
+    if count > theMax then
+      theMax = count;
+    theAvg += count;
+  }
+  theAvg = round(theAvg * 1.0 / histo.size):int;
+  return (theMin, theAvg, theMax);
+}
+
+proc check(name:string, histo, alwaysPrintAvg:bool = false) {
+  var (theMin, theAvg, theMax) = summarize(histo);
+  if verbose then
+    writef("% 32s  min=% 8i  avg=% 8i  max=% 8i\n",
+           name, theMin, theAvg, theMax);
+  else if alwaysPrintAvg then
+    writef("% 32s  avg=% 8i\n", name, theAvg);
+
+  // This comparison is simply a heuristic and is not
+  // based on concrete statistics; just trying to get
+  // the test to fail if the statistics become really bad.
+  if theMax > 10 && theMax > 4*theAvg then
+    writeln(" SPATIAL ANOMALY DETECTED in test ", name);
+}
+
+proc runTest() {
+  
+  var seeds : [1..3] uint(64);
+
+  var topBits = 1 + log2(N);
+  var botBits = topBits;
+  assert(topBits < 32);
+  var nTops = 1 << topBits;
+  var nBots = 1 << botBits;
+  var topCounts:[0..nTops] int;
+  var botCounts:[0..nBots] int;
+
+  var counts0: [0..#M] int;
+  var counts1: [0..#M] int;
+  var counts2: [0..#M] int;
+  var counts3: [0..#M] int;
+
+  // pairs[j,k] = # of collisions of 2 hashes
+  var pairs : [0..#M, 0..#M] int;
+  var pairsInput: [0..#M, 0..#M] list(int);
+  // triples[j,k,l] = # of collisions of 3 hashes
+  var triples : [0..#M, 0..#M, 0..#M] int(64);
+
+  var hashSet:set(uint);
+
+  if useSHA || useSHASeeds {
+    for i in 1..3 {
+      seeds[i] = sha256hash(i:uint);
+    }
+  } else {
+    var rng = new RandomStream(uint(64), seed=0x1BAD5EED);
+    rng.fillRandom(seeds);
+  }
+
+  for i in 1..N {
+    var hash = doHash(i);
+    counts0[shrink(hash)] += 1;
+    if hashSet.contains(hash) {
+      // Error
+      writeln(" SPATIAL ANOMIALY DETECTED: hashSet contains duplicate");
```

Comment:
So close...

Reply:
Are you saying you want me to spell ANOMALY correctly?